Automattic/php-push
# Add end-to-end-test for APNSNetworkService

Adds end-to-end tests for APNSNetworkService.

Uses a small Node.js script to fire up the HTTP/2 server and generate responses. Right now it only tests resuming following a `GOAWAY` frame, which is the most difficult failure mode to simulate. In the future, we could add special-case URL patterns to test other functionality.

**To Test:**
1. Grab this patch, and get set up with a node.js toolchain on your local machine (I'd recommend `brew install nvm`, then in `tests/MockAPNSServer`, run `nvm use` (then follow the directions if needed) to get the right toolchain running.
2. From the `tests/MockAPNSServer` directory, run `yarn setup && yarn install && yarn serve`.
3. In a different terminal window, switch back to the root directory and run `composer test`.

All tests should pass both locally and in CI – it's normal to see `Request failed: 16` – that's a failing request that's part of the test suite.
